### DISCUSSION TITLE
Voltage Select Testing Typo?

### DISCUSSION THREAD

**Discussion by Krystman** [sanni/cartreader#894] [Need help]
I just installed the Voltage Select chip. There is a [section in the Wiki](https://github.com/sanni/cartreader/wiki/Automatic-Voltage-Selection) about testing it. I think there might be a mistake. Here is what it says

<img width="319" alt="Screenshot 2024-01-28 095915" src="https://github.com/sanni/cartreader/assets/10092987/d082f85e-1ce4-430b-9111-427a1abbae72">

According to this pin 8 should change voltage. But that is not what is happening for me. Looking at the pinout that's also not what should be happening. Pin 8 is an input and is wired to 3.3v. Pin 7 is the output. I think this is the one that should change voltage depending on the system. This is what is happening in my build.

I think this is a typo in the Wiki. But before I do changes, I wanted to verify with somebody more familiar with the hardware.

Also I think it would be a good idea to present the pinout of the TPS2113 IC upside-down in this section because that is how it is installed on the board. Took me a while to find out that I have been testing the wrong pins at first.

**Comment by mrfitzy** [comment]
I just installed Vselect, and I am seeing the same results: Pin 7 changes voltage, while Pins 6,8 are 5V,3.3V respectively.

I wouldn't say I'm more familiar with the hardware, but this makes two of us :) I'd say go ahead and fix the wiki.

I think it would also be good to update the wiki steps for testing 5V to specifically **read** a 5V cart (without anything in the slot) before probing. Right now it says to "select Mega Drive/Genesis", but this is a bit vague, and the firmware doesn't generally select the voltage at this point in the menus (same with Game Boy). You need to try to read a cart or something similar.

**Reply by Ancyker** [reply]
The voltage used to switch on the menu. @sanni changed it, probably because if you selected a menu option by accident you might send 5V to your 3.3V cart. 

I'm preparing an update to the firmware which allows you to switch to 5V in a more intuitive "no cart should be inserted" way. When powering on the OSCR without the SD card while Vselect is enabled it will display the usual error with the addition of prompting you to press the button to switch to 5V. This is useful for both testing and flashing the Arduino, which should be done at 5V.

![image](https://github.com/sanni/cartreader/assets/7105333/845cbf3b-6eec-4471-acc0-164dc3222ea8)

For safety, it doesn't immediately switch to 5V, just in case you have a 3.3V cart inserted and just forgot to put the SD card back into the OSCR or something. It simply tells you to press the button to enable 5V.

I'm also adding a way to control the voltage over serial with `SETVOLTS` (a `GETVOLTS` command is already implemented).

The shift to using the SD card being removed to enable 5V was chosen because of the Firmware Updater program I'm working on. Originally I planned to use the `SETVOLTS` command but for it to update the files on the SD card you need to remove it from the OSCR and insert it into your computer's SD card slot. Since you had to do this anyway I figured I might as well have it also switch to 5V on that menu rather than having the updater send a command.

The upcoming revision of the main PCB allows placing a 3.3V regulator as well as all of the other components (clock gen and Arduino) onto the main PCB instead of on modules. Because the Arduino system doesn't use all of the pins of the ATMEGA2560, placing it directly on the board gave me access to those pins. As such, I was able to add a few features to the board and firmware that I was not able to prior because of the lack of IO pins.

When installing (soldering) the ATMEGA2560 (and supporting components) directly on the main PCB (requires the use of a hot plate or hot air + solder paste), it enables the following features:
* **A dedicated RGB LED for VSELECT**. This indicates the status of VSELECT. There are several statuses that can happen at first:
  * Blue: 5V
  * Green: 3.3V
  * Yellow: Low voltage on either 5V or 3.3V.
  * Red: Error, something is wrong, do use the OSCR when the light is red.
* **A voltage sensor to detect what voltage VCC actually is.** This is what can cause the LED mentioned above to turn red. After the firmware requests a voltage it will then check that it's getting that voltage. If it isn't, it will set the LED to red.
* **Voltage sensors to verify that 5V and 3.3V are correct.** If either 5V or 3.3V is low, the firmware will set the LED to yellow.

The following features are also coming with the new revision and can be used even if you aren't installing the ATMEGA2560 directly on the PCB (note that all of the components for these are SMT, the ones that require hot air or a hot plate are noted):
* **Status LEDs for the voltage rails.** These LEDs are connected directly to VBUS (USB), 5V (which is just VBUS but after the power switch), and 3.3V. That means as soon as USB is connected the VBUS LED will light indicating the USB port has power. When you switch the OSCR on the other 2 LEDs should also illuminate. It may be difficult to hand-solder the LED without damaging it.
* **VBUS (USB) surge protection + noise filtering.** SMT footprints are available to add additional protection to the VBUS power rail. There are several components you can install as desired. For instance, if you only want noise filtering you can just install the ferrite bead, which is large enough (0805) that most people shouldn't have much of an issue hand-soldering it. If you only want surge protection, you can just install that IC and nothing else. Note that the surge protection IC *requires* the use of either a hot plate or hot air, it cannot be hand-soldered ([DFN package](https://en.wikipedia.org/wiki/Flat_no-leads_package) with a thermal pad -- the thermal pad *must* be connected or it won't be able to provide proper protection, so it's genuinely impossible to install this by hand soldering it).
* **On-board USB-C.** Components are available to allow the use of USB-C via the main PCB. This can be done for just power, requiring a USB cable connected to the Arduino module for flashing (DO NOT CONNECT BOTH AT ONCE), or, for both power and data by also installing the USB UART IC. Use of the USB UART IC also requires the installation of several other supporting components, which includes a few resistors and a USB Type-C data-line protection IC. The protection IC protects the data lines from being shorted to VBUS. It along with some resistors also allows the use of "real" Type-C to Type-C (Power Delivery) chargers. It is configured to request up to 5V 3A. It's a bit overkill, but the OSCR does use around 250-300mA depending on what cart is connected. This can cause the voltage to drop with cheaper supplies, so allowing the use of a good Type-C supply does have a benefit. Of course, the use of a good supply and cable are still important. Installation of the Type-C port requires the use of a hot plate or hot air even for a power-only configuration as the pads for the USB port are underneath it.
* **On-board 3.3V regulator.** You can put a 3.3V regulator directly on the main PCB instead of using 3.3V from the LCD. This allows using HW5 in a headless/serial-based configuration as well as allowing the use of other LCD modules (namely, it should allow the use of the one from BIGTREETECH). There is a jumper that should be opened when using the regulator and closed when using the LCD for 3.3V. For simplicity, the Gerber I will commit to the GitHub repository will default to the jumper being closed. So, if you want to use a regulator and the provided Gerber file you should cut the jumper (technically optional if you haven't closed the 3.3V jumper on the LCD). Alternatively, you can change the footprint of the jumper to be open and export the Gerber yourself.
* **On-board Clock Generator.** The clock generator can be installed directly on the board as well. There isn't much of a point to this for most people as there is no tangible benefit to the OSCR itself by having the clock gen on-board and when purchasing only 1 or 2, the modules are cheaper than the components. Thus, if you are just DIYing a single unit it doesn't make much sense to install the clock gen components on the board, you should stick to the module unless you have another reason outside of cost to put it directly on the board.
* **FPC headers for the connection to the slot adapter PCB.** The footprints for two FPC 40-pin FPC headers have been added to the main PCB. These provide an alternative way to connect to the adapter PCB vs using the pin headers. Currently, only the seven-slot adapter is planned to support this method.
* **Might also have an FPC header for the LCD.** The LCD from BIGTREETECH has an FPC header as a connection option and I'm currently trying to find a way to fit this onto the board (so, it might not happen, it's getting tight on space). Alone, this wouldn't be a big deal, especially since the BIGTREETECH LCDs do not have the option to output 3.3V. However, with the other changes above the LCD headers are one of the tallest things left on the board, especially when a cable is connected to them (the only other tall thing left is the bulk cap). By eliminating them and tweaking what bulk cap you use, it would be possible to make a very thin OSCR.
* **ISP/ICSP header for the ATMEGA2560.** A programming header has been added to allow burning the bootloader onto the mega. Although it was primarily added as a method to flash a mega that was installed directly on the PCB, it can be used to burn one to a module as well. If you like to enable support for a lot of different systems and find yourself having to disable one or two it might be worth using this header to also flash the firmware. When flashing the firmware using the ISP/ICSP header it will omit the bootloader. That frees up some PROGMEM (~6-8KB) which could allow you to squeeze an extra system or two into the firmware. However, without the bootloader, you will always need to update the firmware using an external programmer. If you are using this method you can cut the JP201 (Reset Enable) jumper so that the Arduino isn't reset when the serial console is connected as this is only used for flashing the firmware, though it isn't necessary to do it.

![image](https://github.com/sanni/cartreader/assets/7105333/0e25a10f-b895-4da8-8c26-0c9eceeaa32f)

Well, that was a lot longer than I intended/expected... :|

**Reply by mrfitzy** [reply]
Very much looking forward to having on-board USB-C when this becomes available :)

**Comment by Ancyker** [comment]
I corrected it.

**Comment by Krystman** [comment]
Great! I also updated the diagram

![TPS2113](https://github.com/sanni/cartreader/assets/10092987/384560fd-c0c4-4000-96f8-39e2dc0e6060)

**Reply by Ancyker** [reply]
`STAT` can be used on REV6 so shouldn't be left unlabeled.

![image](https://github.com/sanni/cartreader/assets/7105333/ae37f08f-b4f9-477f-a518-a5bc75a9dd0e)

**Reply by Ancyker** [reply]
`STAT` connects to the physical pin PE6 of the on-board ATMEGA2560, when installed.

![image](https://github.com/sanni/cartreader/assets/7105333/193f4c07-4057-4b3f-ab14-0391b2b2d8e2)

![image](https://github.com/sanni/cartreader/assets/7105333/1a3516b1-a2c5-4804-af0b-7f142752ebd3)

**Reply by Krystman** [reply]
Is pin 1 routed anywhere in REV5?

**Reply by Ancyker** [reply]
No, only in REV6. There are no free IO pins on the module we are using so there was nothing to connect it to. The Arduino project didn't use all of the pins of the ATMEGA2560, so by placing it directly on the board there are more IO pins accessible.

**Reply by Krystman** [reply]
Ok, updated
![TPS2113](https://github.com/sanni/cartreader/assets/10092987/bf870c78-27aa-41c8-bdb8-81e6e0728443)
